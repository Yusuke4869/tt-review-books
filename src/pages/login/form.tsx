import { useState } from "react";
import { useCookies } from "react-cookie";
import { useForm } from "react-hook-form";
import { useLocation } from "wouter";

import { useAuth } from "~/contexts/auth";

import type { FC } from "react";
import type { SubmitHandler } from "react-hook-form";

import type { APIError, SignInResponse } from "~/types/api";

type Inputs = {
  email: string;
  password: string;
};

export const LoginForm: FC = () => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<Inputs>();
  const [, setCookie] = useCookies();
  const [, navigate] = useLocation();
  const [errorMessage, setErrorMessage] = useState<string | null>(null);

  const { setAuthenticated } = useAuth();

  const onSubmit: SubmitHandler<Inputs> = async (data) => {
    const res = await fetch(`${import.meta.env.VITE_API_URL}/signin`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        email: data.email,
        password: data.password,
      }),
    });

    if (!res.ok) {
      const r = (await res.json()) as APIError;
      setErrorMessage(r.ErrorMessageJP);
      return;
    }

    const r = (await res.json()) as SignInResponse;
    setCookie("token", r.token);
    setAuthenticated(true);
    navigate("/");
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div className="flex flex-col gap-y-2 mb-3 w-3/5 lg:w-2/5 xl:w-1/4">
        <div className="flex flex-col gap-y-0.5">
          <label>メールアドレス</label>
          <input
            type="text"
            className="rounded-md border-2 border-gray-300 outline-none text-gray-900 p-1.5 focus:border-indigo-600"
            {...register("email", {
              required: true,
              pattern: /^[a-zA-Z0-9_+-]+(.[a-zA-Z0-9_+-]+)*@([a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]*\.)+[a-zA-Z]{2,}$/,
            })}
          />
          {errors.email && (
            <p className="text-red-500 text-sm" data-error-message="email">
              正しいメールアドレスを入力してください
            </p>
          )}
        </div>
        <div className="flex flex-col gap-y-0.5">
          <label>パスワード</label>
          <input
            type="password"
            className="rounded-md border-2 border-gray-300 outline-none text-gray-900 p-1.5 focus:border-indigo-600"
            {...register("password", { required: true })}
          />
          {errors.password && (
            <p className="text-red-500 text-sm" data-error-message="password">
              パスワードを入力してください
            </p>
          )}
        </div>
      </div>
      {errorMessage && <p className="text-red-500 text-sm my-4">{errorMessage}</p>}
      <input
        type="submit"
        className="cursor-pointer text-white bg-blue-700 hover:bg-blue-800 rounded-lg text-sm px-5 py-2.5 focus:outline-none"
        value="ログイン"
      />
    </form>
  );
};
