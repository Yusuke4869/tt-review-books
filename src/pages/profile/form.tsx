import Compressor from "compressorjs";
import { useEffect, useState } from "react";
import { useCookies } from "react-cookie";
import { useForm } from "react-hook-form";

import type { ChangeEvent, FC } from "react";
import type { SubmitHandler } from "react-hook-form";

import type { APIError, UpdateUserResponse, UploadUserIconResponse, User } from "~/types/api";

type Props = {
  user: User;
  setUser: (user: User) => void;
};

type Inputs = {
  name: string;
  icon: FileList;
};

export const ProfileForm: FC<Props> = ({ user, setUser }) => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<Inputs>({
    defaultValues: {
      name: user.name,
    },
  });
  const [cookie] = useCookies(["token"]);
  const [image, setImage] = useState<string | null>(null);
  const [errorMessage, setErrorMessage] = useState<string | null>(null);

  useEffect(() => {
    if (user.iconUrl) setImage(user.iconUrl);
  }, [user.iconUrl]);

  const onSubmit: SubmitHandler<Inputs> = async (data) => {
    const res = await fetch(`${import.meta.env.VITE_API_URL}/users`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${cookie.token}`,
      },
      body: JSON.stringify({
        name: data.name,
      }),
    });

    if (!res.ok) {
      const r = (await res.json()) as APIError;
      setErrorMessage(r.ErrorMessageJP);
      return;
    }

    const r = (await res.json()) as UpdateUserResponse;
    setUser({ ...user, name: r.name });

    const icon = data.icon[0];
    if (!icon) return;
    new Compressor(icon, {
      quality: 0,
      success(result) {
        const formData = new FormData();
        formData.append("icon", result, icon.name);

        fetch(`${import.meta.env.VITE_API_URL}/uploads`, {
          method: "POST",
          headers: {
            Authorization: `Bearer ${cookie.token}`,
          },
          body: formData,
        }).then(async (res) => {
          if (res.ok) {
            const r = (await res.json()) as UploadUserIconResponse;
            setUser({ ...user, iconUrl: r.iconUrl });
          }
        });
      },
    });
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div className="flex flex-col gap-y-2 mb-3 w-3/5 lg:w-2/5 xl:w-1/4">
        <div className="flex flex-col gap-y-0.5">
          <label htmlFor="name">ユーザー名</label>
          <input
            id="name"
            type="text"
            className="rounded-md border-2 border-gray-300 outline-none text-gray-900 p-1.5 focus:border-indigo-600"
            {...register("name", {
              required: true,
            })}
          />
          {errors.name && <p className="text-red-500 text-sm">ユーザー名を入力してください</p>}
        </div>
        <div className="flex flex-col gap-y-0.5">
          <label>アイコン</label>
          {image && !errors.icon && <img src={image} className="w-20 h-20" />}
          <input
            type="file"
            className="rounded-md outline-none text-gray-900 p-1.5"
            {...register("icon", {
              onChange(e: ChangeEvent<HTMLInputElement>) {
                const files = e.target.files;
                if (!files || files.length <= 0) return;
                const reader = new FileReader();
                reader.onload = () => setImage(reader.result as string);
                reader.readAsDataURL(files[0]);
              },
            })}
            accept="image/png, image/jpeg"
          />
        </div>
      </div>
      {errorMessage && <p className="text-red-500 text-sm my-4">{errorMessage}</p>}
      <input
        type="submit"
        className="cursor-pointer text-white bg-blue-700 hover:bg-blue-800 rounded-lg text-sm px-5 py-2.5 focus:outline-none"
        value="登録情報更新"
      />
    </form>
  );
};
