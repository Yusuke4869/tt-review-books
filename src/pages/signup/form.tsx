import Compressor from "compressorjs";
import { useState } from "react";
import { useCookies } from "react-cookie";
import { useForm } from "react-hook-form";
import { useLocation } from "wouter";

import { useAuth } from "~/contexts/auth";

import type { ChangeEvent, FC } from "react";
import type { SubmitHandler } from "react-hook-form";

import type { APIError, CreateUserResponse } from "~/types/api";

type Inputs = {
  name: string;
  email: string;
  password: string;
  icon: FileList;
};

export const SignUpForm: FC = () => {
  const {
    register,
    handleSubmit,
    formState: { errors },
  } = useForm<Inputs>();
  const [, setCookie] = useCookies();
  const [, navigate] = useLocation();
  const [image, setImage] = useState<string | null>(null);
  const [errorMessage, setErrorMessage] = useState<string | null>(null);

  const { setAuthenticated } = useAuth();

  const onSubmit: SubmitHandler<Inputs> = async (data) => {
    const res = await fetch(`${import.meta.env.VITE_API_URL}/users`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        name: data.name,
        email: data.email,
        password: data.password,
      }),
    });

    if (!res.ok) {
      const r = (await res.json()) as APIError;
      setErrorMessage(r.ErrorMessageJP);
      return;
    }

    const r = (await res.json()) as CreateUserResponse;

    const icon = data.icon[0];
    if (!icon) return;
    new Compressor(icon, {
      quality: 0,
      success(result) {
        const formData = new FormData();
        formData.append("icon", result, icon.name);

        fetch(`${import.meta.env.VITE_API_URL}/uploads`, {
          method: "POST",
          headers: {
            Authorization: `Bearer ${r.token}`,
          },
          body: formData,
        });
      },
    });

    setCookie("token", r.token);
    setAuthenticated(true);
    navigate("/");
  };

  return (
    <form onSubmit={handleSubmit(onSubmit)}>
      <div className="flex flex-col gap-y-2 mb-3 w-3/5 lg:w-2/5 xl:w-1/4">
        <div className="flex flex-col gap-y-0.5">
          <label htmlFor="name">ユーザー名</label>
          <input
            id="name"
            type="text"
            className="rounded-md border-2 border-gray-300 outline-none text-gray-900 p-1.5 focus:border-indigo-600"
            {...register("name", {
              required: true,
              pattern: /^[a-zA-Z0-9_+-]+(.[a-zA-Z0-9_+-]+)*@([a-zA-Z0-9][a-zA-Z0-9-]*[a-zA-Z0-9]*\.)+[a-zA-Z]{2,}$/,
            })}
          />
          {errors.name && <p className="text-red-500 text-sm">ユーザー名を入力してください</p>}
        </div>
        <div className="flex flex-col gap-y-0.5">
          <label htmlFor="email">メールアドレス</label>
          <input
            id="email"
            type="email"
            className="rounded-md border-2 border-gray-300 outline-none text-gray-900 p-1.5 focus:border-indigo-600"
            {...register("email", { required: true })}
          />
          {errors.email && <p className="text-red-500 text-sm">正しいメールアドレスを入力してください</p>}
        </div>
        <div className="flex flex-col gap-y-0.5">
          <label htmlFor="password">パスワード</label>
          <input
            id="password"
            type="password"
            className="rounded-md border-2 border-gray-300 outline-none text-gray-900 p-1.5 focus:border-indigo-600"
            {...register("password", { required: true })}
          />
          {errors.password && <p className="text-red-500 text-sm">パスワードを入力してください</p>}
        </div>
        <div className="flex flex-col gap-y-0.5">
          <label>アイコン</label>
          {image && !errors.icon && <img src={image} className="w-20 h-20" />}
          <input
            type="file"
            className="rounded-md outline-none text-gray-900 p-1.5"
            {...register("icon", {
              required: true,
              onChange(e: ChangeEvent<HTMLInputElement>) {
                const files = e.target.files;
                if (!files || files.length <= 0) return;
                const reader = new FileReader();
                reader.onload = () => setImage(reader.result as string);
                reader.readAsDataURL(files[0]);
              },
            })}
            accept="image/png, image/jpeg"
          />
          {errors.icon && <p className="text-red-500 text-sm">画像を選択してください</p>}
        </div>
      </div>
      {errorMessage && <p className="text-red-500 text-sm my-4">{errorMessage}</p>}
      <input
        type="submit"
        className="cursor-pointer text-white bg-blue-700 hover:bg-blue-800 rounded-lg text-sm px-5 py-2.5 focus:outline-none"
        value="ユーザー登録"
      />
    </form>
  );
};
