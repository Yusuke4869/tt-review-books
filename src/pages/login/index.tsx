import { Link, Redirect } from "wouter";

import { Layout } from "~/components/layout";
import { useAuth } from "~/contexts/auth";
import { LoginForm } from "./form";

import type { FC } from "react";

export const Login: FC = () => {
  const { authenticated } = useAuth();
  if (authenticated) return <Redirect to="/" />;

  return (
    <Layout>
      <div className="mx-auto ml-20">
        <h2 className="text-xl font-semibold mt-6 mb-3">ログイン</h2>
        <LoginForm />
        <Link href="/signup" className="inline-block text-blue-700 underline my-3">
          アカウントをお持ちでない方は新規登録
        </Link>
      </div>
    </Layout>
  );
};
