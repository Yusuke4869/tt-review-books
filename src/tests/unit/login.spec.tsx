import "@testing-library/jest-dom";
import { render } from "@testing-library/react";
import { expect, test } from "vitest";

import { LoginForm } from "~/pages/login/form";

test("email入力欄が表示される", () => {
  const { getByLabelText } = render(<LoginForm />);
  expect(getByLabelText("メールアドレス")).toBeInTheDocument();
  expect(getByLabelText("メールアドレス")).toHaveAttribute("type", "email");
});

test("password入力欄が表示される", () => {
  const { getByLabelText } = render(<LoginForm />);
  expect(getByLabelText("パスワード")).toBeInTheDocument();
  expect(getByLabelText("パスワード")).toHaveAttribute("type", "password");
});

test("ログインボタンが表示される", () => {
  const { getByText } = render(<LoginForm />);
  expect(getByText("ログイン")).toBeInTheDocument();
  expect(getByText("ログイン")).toHaveAttribute("type", "submit");
});

test("スナップショットと一致する", () => {
  const { container } = render(<LoginForm />);
  expect(container).toMatchSnapshot();
});
