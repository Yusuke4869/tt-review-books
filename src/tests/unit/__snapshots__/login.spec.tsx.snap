// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`スナップショットと一致する 1`] = `
<div>
  <form>
    <div
      class="flex flex-col gap-y-2 mb-3 w-3/5 lg:w-2/5 xl:w-1/4"
    >
      <div
        class="flex flex-col gap-y-0.5"
      >
        <label
          for="email"
        >
          メールアドレス
        </label>
        <input
          class="rounded-md border-2 border-gray-300 outline-none text-gray-900 p-1.5 focus:border-indigo-600"
          id="email"
          name="email"
          type="email"
        />
      </div>
      <div
        class="flex flex-col gap-y-0.5"
      >
        <label
          for="password"
        >
          パスワード
        </label>
        <input
          class="rounded-md border-2 border-gray-300 outline-none text-gray-900 p-1.5 focus:border-indigo-600"
          id="password"
          name="password"
          type="password"
        />
      </div>
    </div>
    <input
      class="cursor-pointer text-white bg-blue-700 hover:bg-blue-800 rounded-lg text-sm px-5 py-2.5 focus:outline-none"
      type="submit"
      value="ログイン"
    />
  </form>
</div>
`;
